tools/src/bin/build-blame.rs

```diff
@@ -0,0 +1,357 @@
+extern crate env_logger;
+extern crate git2;
+#[macro_use]
+extern crate log;
+extern crate tools;
+extern crate unicode_normalization;
+
+use std::borrow::Borrow;
+use std::collections::HashMap;
+use std::env;
+use std::path::{Path, PathBuf};
+use std::process::Command;
+
+use git2::{DiffFindOptions, ObjectType, Oid, Patch, Repository, Sort};
+use tools::config::index_blame;
+use tools::git_ops;
+use unicode_normalization::UnicodeNormalization;
+
+fn get_hg_rev(git_repo: &Repository, git_oid: &Oid) -> Option<String> {
+    // TODO: start a helper process and stream I/O to it instead of spawning a new process every time
+    let output = Command::new("git")
+        .arg("cinnabar")
+        .arg("git2hg")
+        .arg(format!("{}", git_oid))
+        .current_dir(git_repo.path())
+        .output()
+        .unwrap();
+    if !output.status.success() {
+        panic!("`git cinnabar git2hg {}` returned error code {:?}", git_oid, output.status.code());
+    }
+    match git_ops::decode_bytes(output.stdout).lines().next() {
+        Some(r) if r != "0000000000000000000000000000000000000000" => Some(r.to_string()),
+        _ => None,
+    }
+}
+
+fn count_lines(blob: &git2::Blob) -> usize {
+    let data = blob.content();
+    if data.len() == 0 {
+        return 0;
+    }
+    let mut linecount = 0;
+    for b in data {
+        if *b == b'\n' {
+            linecount += 1;
+        }
+    }
+    if data[data.len() - 1] != b'\n' {
+        linecount += 1;
+    }
+    linecount
+}
+
+fn unmodified_lines(
+    blob: &git2::Blob,
+    parent_blob: &git2::Blob,
+) -> Result<Vec<(usize, usize)>, git2::Error> {
+    let mut unchanged = Vec::new();
+
+    let patch = Patch::from_blobs(
+        parent_blob,
+        None,
+        blob,
+        None,
+        None,
+    )?;
+
+    if patch.delta().flags().is_binary() {
+        return Ok(unchanged);
+    }
+
+    fn add_delta(lineno: usize, delta: i32) -> usize {
+        ((lineno as i32) + delta) as usize
+    }
+
+    let mut latest_line : usize = 0;
```

Comment:
nit: Remove the space to the left of `:`? That seems more common, and it is consistent with arguments.

Reply:
Fixed